src/storage/src/source/persist_source.rs

```diff
@@ -190,3 +194,218 @@ where
+            .expect("could not open persist client")
+            .clone();
+
+        let mut read = persist_client
+            .open_reader::<SourceData, (), mz_repr::Timestamp, mz_repr::Diff>(data_shard.clone())
+            .await
+            .expect("could not open persist shard");
+
+        // Aggressively downgrade `since`, to not hold back compaction.
+        read.downgrade_since(as_of_moveable.clone()).await;
+
+        let snapshots_hollow_batches = read
+            .snapshot_hollow_batch(as_of_moveable.clone())
+            .await
+            .expect("cannot serve requested as_of");
+
+        // First, yield all the updates from the snapshot.
+        for batch in snapshots_hollow_batches {
+            yield batch;
+        }
+
+        // Then, listen continously and yield any new updates. This loop is expected to never
+        // finish.
+        let mut listen = read
+            .listen(as_of_moveable)
+            .await
+            .expect("cannot serve requested as_of");
+
+        loop {
+            yield listen.listen_hollow_batch().await;
+        }
+    });
+
+    let mut pinned_stream = Box::pin(async_stream);
+
+    let (inner, token) =
+        crate::source::util::source(scope, "persist_source".to_string(), move |info| {
+            let waker_activator = Arc::new(scope.sync_activator_for(&info.address[..]));
+            let waker = futures_util::task::waker(waker_activator);
+            let activator = scope.activator_for(&info.address[..]);
+
+            // There is a bit of a mismatch: listening on a ReadHandle will give us an Antichain<T>
+            // as a frontier in `Progress` messages while a timely source usually only has a single
+            // `Capability` that it can downgrade. We can work around that by using a
+            // `CapabilitySet`, which allows downgrading the contained capabilities to a frontier
+            // but `util::source` gives us both a vanilla `Capability` and a `CapabilitySet` that
+            // we have to keep downgrading because we cannot simply drop the one that we don't
+            // need.
+
+            let mut current_ts = 0;
+
+            move |cap, output| {
+                let mut context = Context::from_waker(&waker);
+                // Bound execution of operator to prevent a single operator from hogging
+                // the CPU if there are many messages to process
+                let timer = Instant::now();
+
+                let mut i = 0;
+
+                while let Poll::Ready(item) = pinned_stream.as_mut().poll_next(&mut context) {
+                    match item {
+                        Some(Ok(batch)) => {
+                            let session_cap = cap.delayed(&current_ts);
+                            let mut session = output.session(&session_cap);
+                            let progress = batch.generate_progress();
+
+                            session.give((i, batch));
+
+                            // Round robin
```

Comment:
round robin is what I was thinking, too! :clap:

Reply:
Minorest of nits: round robin occasionally gets itself in trouble with self-synchonization: you have a few sources running, and they all use the same strategy, and .. surprisingly often they end up working in lock step. So, if you have 10 sources, each time slice one worker gets 10 batches rather than distributing them (the round robin counter isn't shared across sources). Easy-ish fix would be to use `(source_id, i).hashed() % peers` or something like it!

Reply:
Implemented Frank's suggestions here.